class-fields-registry.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+final class Fields_Registry {
+	protected $datastores = [];
+
+	protected $datastore_types = [];
+
+	protected $control_types = [];
+
+	protected static $instance = null;
+	protected        $controls = [];
+
+	/**
+	 * @return $this
+	 */
+	public static function get_instance() : Fields_Registry {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new self;
+		}
+
+		return self::$instance;
+	}
+
+	public function register_datastore_type( string $type, string $instance ) {
+		if ( ! isset( $this->datastore_types[ $type ] ) ) {
+			$this->datastore_types[ $type ] = $instance;
+		}
+
+		return $this;
+	}
+
+	public function register_control_type( string $type, string $instance ) {
+		if ( ! isset( $this->control_types[ $type ] ) ) {
+			$this->control_types[ $type ] = $instance;
+		}
+
+		return $this;
+	}
+
+	/**
+	 * @param string $type
+	 *
+	 * @return string|null
+	 */
+	protected function get_field_control_type( string $type ) {
+		return isset( $this->control_types[ $type ] ) && is_string( $this->control_types[ $type ] ) && is_subclass_of( $this->control_types[ $type ], Field_Control::class ) ? $this->control_types[ $type ] : null;
+	}
+
+	/**
+	 * @param string $type
+	 *
+	 * @return string|null
+	 */
+	protected function get_field_datastore_type( string $type ) {
+		return isset( $this->datastore_types[ $type ] ) && is_string( $this->datastore_types[ $type ] ) && is_subclass_of( $this->datastore_types[ $type ], Field_Data_Store::class ) ? $this->datastore_types[ $type ] : null;
+	}
+
+	public function register_json( string $file ) {
+		$fields = json_decode( file_get_contents( $file ), true );
+
+		foreach ( $fields as $field ) {
+			if ( isset( $field['datastore'] ) && is_array( $field['datastore'] ) && isset( $field['controls'] ) && is_array( $field['controls'] ) ) {
+				$id             = isset( $field['id'] ) && is_string( $field['id'] ) ? $field['id'] : null;
+				$datastore      = $this->get_field_datastore_type( $field['datastore']['type'] ?? '' );
+				$datastore_args = $field['datastore'];
+
+				if ( $id && $datastore ) {
+					$this->register_datastore( $id, $datastore, $datastore_args );
+				}
+
+				foreach ( $field['controls'] as $control_id => $control_data ) {
+					$control    = $this->get_field_control_type( $control_data['type'] ?? '' );
+
+					if ( $id && $control && $control_id ) {
+						$this->register_control( $id, $control_id, $control, $control_data );
+					}
+				}
+			}
+		}
+	}
+
+	protected function get_field_datastore_instance( string $id ) {
+		$type = $this->datastores[ $id ]['type'];
```

Comment:
I understand usage of this method currently checks `if ( isset( $this->datastores[ $id ] ) )` but should that be in this method instead or both?